### PR TITLE
Service Bus Named Key Credential

### PR DESCRIPTION
**Pull request by rakshith91**
Fixes https://github.com/Azure/azure-sdk-for-python/issues/17905

### CODE REVIEW

File: sdk/servicebus/azure-servicebus/azure/servicebus/_base_handler.py [hunk 1]
```diff
@@ -196,6 +196,22 @@ def get_token(self, *scopes, **kwargs):  # pylint:disable=unused-argument
...
+    :param str credential: The AzureNamedKeyCredential that should be used
+    """
+
+    def __init__(self, credential):
+        # type: (AzureNamedKeyCredential) -> None
+        self.credential = credential
+        self.token_type = b"servicebus.windows.net:sastoken"
+
+    def get_token(self, *scopes, **kwargs):  # pylint:disable=unused-argument
+        # type: (str, Any) -> _AccessToken
```
Comment (rakshith91) [inline sdk/servicebus/azure-servicebus/azure/servicebus/_base_handler.py:210]:
```suggestion
        # type: (str, Any) -> AccessToken
```

File: sdk/servicebus/azure-servicebus/azure/servicebus/_base_handler.py [hunk 2]
```diff
@@ -196,6 +196,22 @@ def get_token(self, *scopes, **kwargs):  # pylint:disable=unused-argument
             raise ValueError("No token scope provided.")
         return _generate_sas_token(scopes[0], self.policy, self.key)
 
+class AzureNamedKeyTokenCredential(object):
```
Comment (yunhaoling) [inline sdk/servicebus/azure-servicebus/azure/servicebus/_base_handler.py:199]: same comment as in the [eventhub PR](https://github.com/Azure/azure-sdk-for-python/pull/18292), rename to something like:
```suggestion
class ServiceBusAzureNamedKeyCredential(object):
```

File: sdk/servicebus/azure-servicebus/azure/servicebus/_base_handler.py [hunk 3]
```diff
@@ -196,6 +196,22 @@ def get_token(self, *scopes, **kwargs):  # pylint:disable=unused-argument
...
+    """The named key credential used for authentication.
+    :param str credential: The AzureNamedKeyCredential that should be used
```
Comment (yunhaoling) [inline sdk/servicebus/azure-servicebus/azure/servicebus/_base_handler.py:201]: not a str
```suggestion
    :param credential: The AzureNamedKeyCredential that should be used.
    :type credential: ~azure.core.credentials.AzureNamedKeyCredential
```

File: sdk/servicebus/azure-servicebus/azure/servicebus/_base_handler.py [hunk 4]
```diff
@@ -196,6 +196,22 @@ def get_token(self, *scopes, **kwargs):  # pylint:disable=unused-argument
...
+        # type: (AzureNamedKeyCredential) -> None
+        self.credential = credential
+        self.token_type = b"servicebus.windows.net:sastoken"
+
+    def get_token(self, *scopes, **kwargs):  # pylint:disable=unused-argument
+        # type: (str, Any) -> AccessToken
+        if not scopes:
+            raise ValueError("No token scope provided.")
+        name, key = self.credential.named_key
+        return _generate_sas_token(scopes[0], name, key)
```
Comment (yunhaoling) [inline sdk/servicebus/azure-servicebus/azure/servicebus/_base_handler.py:214]: same code style, see comment in the eventhub PR: https://github.com/Azure/azure-sdk-for-python/pull/18292#discussion_r625101101

**Comment by yunhaoling** [comment]
thanks for PR! I think most of the comments I left in the [eventhub named key one](https://github.com/Azure/azure-sdk-for-python/pull/18292) would apply to here as well.

also samples are missing 😉

File: sdk/servicebus/azure-servicebus/samples/async_samples/authenticate_using_named_key_credential.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
+#!/usr/bin/env python
+
+# --------------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
+"""
+An example to show authentication using AzureNamedKeyCredential.
```
Comment (yunhaoling) [inline sdk/servicebus/azure-servicebus/samples/async_samples/authenticate_using_named_key_credential.py:8]: nit: the file name should have `_async` suffix.

**Review by yunhaoling — APPROVED** [review]

**Pull request merged**